pkg/cloudprovider/providers/gce/gce_firewall.go

```diff
@@ -112,3 +115,33 @@ func (gce *GCECloud) UpdateFirewall(name, desc string, sourceRanges netsets.IPNe
 
 	return mc.Observe(gce.updateFirewall(name, region, desc, sourceRanges, svcPorts, hosts))
 }
+
+// packProtocolsPortsFromFirewallAlloweds packs protocols and ports from FirewallAlloweds
+// in an unified way for verification.
+func packProtocolsPortsFromFirewallAlloweds(alloweds []*compute.FirewallAllowed) []string {
+	protocolPorts := []string{}
+	for _, allowed := range alloweds {
+		for _, port := range allowed.Ports {
+			protocolPorts = append(protocolPorts, strings.ToLower(allowed.IPProtocol+":"+port))
+		}
+	}
+	return protocolPorts
+}
+
+// packProtocolsPortsFromServicePorts packs protocols and ports from ServicePorts
+// in an unified way for verification.
+func packProtocolsPortsFromServicePorts(svcPorts []v1.ServicePort) []string {
+	protocolPorts := []string{}
+	for _, sp := range svcPorts {
+		protocolPorts = append(protocolPorts, strings.ToLower(fmt.Sprintf("%v:%v", sp.Protocol, sp.Port)))
+	}
+	return protocolPorts
+}
+
+// sameStringArray verifies whether two string arrays have the same strings, order
+// does not matter.
+func sameStringArray(result, expected []string) bool {
+	res := sets.NewString(result...)
+	exp := sets.NewString(expected...)
+	return res.Equal(exp)
+}
```

Comment:
Found we already have `slicesEqual()` in gce_loadbalancer.go. Removing this function and moving that to gce_utils.